java_api/file_api/src/main/java/uk/ramp/metadata/MetadataItem.java

```diff
@@ -100,57 +108,94 @@ default MetadataItem applyOverrides(List<OverrideItem> overrides) {
 
     MetadataItem overriddenMetadataItem = ImmutableMetadataItem.copyOf(this);
     for (MetadataItem override : overridesToApply) {
-      overriddenMetadataItem = applyOverride(overriddenMetadataItem, override);
+      overriddenMetadataItem = applyOverride(overriddenMetadataItem, override, onlyOverrideIfEmpty);
     }
     return overriddenMetadataItem;
   }
 
-  private MetadataItem applyOverride(MetadataItem baseMetadata, MetadataItem metadataOverride) {
+  private boolean shouldOverrideField(
+      Function<MetadataItem, Boolean> isFieldPresent,
+      MetadataItem baseMetadata,
+      MetadataItem overrideMetadata,
+      boolean onlyOverrideIfEmpty) {
+    boolean isBaseFieldPresent = isFieldPresent.apply(baseMetadata);
+    boolean isOverrideFieldPresent = isFieldPresent.apply(overrideMetadata);
+
+    if (!isOverrideFieldPresent) {
+      return false;
+    }
+    if (onlyOverrideIfEmpty) {
+      return !isBaseFieldPresent;
+    }
+    return true;
+  }
+
+  private MetadataItem applyOverride(
+      MetadataItem baseMetadata, MetadataItem metadataOverride, boolean onlyOverrideIfEmpty) {
```

Comment:
what's the scenario in which we don't want to apply overrides?

Reply:
The scenario is when we call open_for_write() and the filename is provided in the metadata params. In this case, we will call applyOverride with onlyOverrideIfEmpty set to True. This will result in the override for generating a fresh filename based on data_product + run_id + extension to not be used if the filename is already provided as metadata in the call to open_for_write()